### DISCUSSION TITLE
Why not visible elements still occupy space in Horizontal/VerticalLayout

### DISCUSSION THREAD

**Discussion by qarmin** [slint-ui/slint#4021] [Q&A]
When element is not visible in Horizontal/Vertical Laout, its space should fill other elements.

Not 100% sure, but I think that this is default GTK and QT behavior - [CODE](https://slint.dev/releases/1.3.0/editor/?snippet=import+%7B+AboutSlint%2C+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++property+%3Cbool%3E+vis+%3A+true%3B%0A++++Rectangle+%7B%0A++++++++background%3A+green%3B%0A++++++++HorizontalLayout+%7B%0A++++++++++++alignment%3A+LayoutAlignment.stretch%3B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++visible%3A+vis%3B%0A++++++++++++++++background%3A+blue%3B%0A++++++++++++++++Text%7B%0A++++++++++++++++++++text%3A+%22A%22%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++visible%3A+vis%3B%0A++++++++++++++++background%3A+blue%3B%0A++++++++++++++++Text%7B%0A++++++++++++++++++++text%3A+%22A%22%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++background%3A+red%3B%0A++++++++++++++++Text%7B%0A++++++++++++++++++++text%3A+%22B%22%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++background%3A+red%3B%0A++++++++++++++++Text%7B%0A++++++++++++++++++++text%3A+%22B%22%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Button+%7B%0A++++++++++++++++text%3A+%22Click+Me%22%3B%0A++++++++++++++++clicked+%3D%3E+%7B%0A++++++++++++++++++++vis+%3D+%21vis%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A&style=native)

https://github.com/slint-ui/slint/assets/41945903/0d4bb8dc-1de7-4632-b0a7-fe2dc7c14877

**Comment by bjorn** [comment]
In a QWidget layout or a QML `RowLayout` / `ColumnLayout`, indeed invisible widgets / items will cause a widget to no longer take up space. However, I think it should be noted that there isn't really any other easy way of implementing a "conditional" widget in these systems.

In contrast, Slint provides an easy way to make a widget conditional. See [this adjusted example code](https://slint.dev/releases/1.3.0/editor/?snippet=import+%7B+AboutSlint%2C+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++property+%3Cbool%3E+vis+%3A+true%3B%0A++++Rectangle+%7B%0A++++++++background%3A+green%3B%0A++++++++HorizontalLayout+%7B%0A++++++++++++alignment%3A+LayoutAlignment.stretch%3B%0A++++++++++++if+%28vis%29%3A+Rectangle+%7B%0A++++++++++++++++background%3A+blue%3B%0A++++++++++++++++Text%7B%0A++++++++++++++++++++text%3A+%22A%22%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++if+%28vis%29%3A+Rectangle+%7B%0A++++++++++++++++background%3A+blue%3B%0A++++++++++++++++Text%7B%0A++++++++++++++++++++text%3A+%22A%22%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++background%3A+red%3B%0A++++++++++++++++Text%7B%0A++++++++++++++++++++text%3A+%22B%22%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++background%3A+red%3B%0A++++++++++++++++Text%7B%0A++++++++++++++++++++text%3A+%22B%22%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++++++Button+%7B%0A++++++++++++++++text%3A+%22Click+Me%22%3B%0A++++++++++++++++clicked+%3D%3E+%7B%0A++++++++++++++++++++vis+%3D+%21vis%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A&style=native).

That said, this changes the behavior, since when the condition changes, it will reset the state of the widget and its children (in case there are any properties). It also makes it so that you can't refer to this widget or its children from outside the conditional block. But I do think it makes it less obvious, what the behavior should be for invisible widgets.

In HTML/CSS, the solution is to not have a boolean "visible" but a "visibility" enum, so it can be set to "visible", "hidden" or "collapse" as desired (and then there is the "display" enum as well of course...).

**Reply by qarmin** [reply]
I considered already conditional widgets, but that would require the use of adding external properties to store the state of the destroyed element to be able to restore it's state when it is re-created.

This complicates/duplicates the code, so I avoid it where possible

**Reply by ogoffart** [reply]
the visible property currently doesn't impact the layout. Maybe it should have, but that would be a breaking change at this point.

Maybe something to think about if we review our layouts in Slint 2.
